### PR TITLE
8295078: TextField blurry when inside an TitledPane -> AnchorPane

### PR DESCRIPTION
**Pull request by Maran23**
The problem here is, that the `AnchorPane` does not use its snapped insets.
Therefore, the fix is to replace all `getInsets().getXXX` calls with their corresponding `snappedXXXInset()` methods.

Note: The reason the `AnchorPane` inside  a `TitledPane` is blurry in the first place is because a `TitledPane` applies padding to its content.
Line 2995 in `modena.css`:
```
.titled-pane > .content > AnchorPane {
    -fx-padding: 0.8em;
}
```
which translates to 9.6px.

EDIT: This is btw a good example of the JUnit 5 feature `@ParameterizedTest` with `@ValueSource`

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8295078](https://bugs.openjdk.org/browse/JDK-8295078): TextField blurry when inside an TitledPane -&gt; AnchorPane

### Reviewers
 * [Ambarish Rapte](https://openjdk.org/census#arapte) (@arapte - **Reviewer**) 🔄 Re-review required (review applies to [9fa887b4](https://git.openjdk.org/jfx/pull/910/files/9fa887b4ceea788ac82ba2120f05a643e633dd2d))
 * [Andy Goryachev](https://openjdk.org/census#angorya) (@andy-goryachev-oracle - **Reviewer**)
 * [Kevin Rushforth](https://openjdk.org/census#kcr) (@kevinrushforth - **Reviewer**)
 * [John Hendrikx](https://openjdk.org/census#jhendrikx) (@hjohn - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/910/head:pull/910` \
`$ git checkout pull/910`

Update a local copy of the PR: \
`$ git checkout pull/910` \
`$ git pull https://git.openjdk.org/jfx.git pull/910/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 910`

View PR using the GUI difftool: \
`$ git pr show -t 910`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/910.diff">https://git.openjdk.org/jfx/pull/910.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/910#issuecomment-1274372884)

### CODE REVIEW

File: modules/javafx.graphics/src/test/java/test/javafx/scene/layout/AnchorPaneTest.java [hunk 1]
```diff
@@ -25,7 +25,14 @@
...
 
-import test.javafx.scene.layout.MockBiased;
+import javafx.beans.property.DoubleProperty;
+import javafx.beans.property.SimpleDoubleProperty;
+import javafx.geometry.Bounds;
+import javafx.scene.Scene;
+import javafx.scene.layout.StackPane;
+import javafx.stage.Stage;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
```
Comment (arapte) [inline modules/javafx.graphics/src/test/java/test/javafx/scene/layout/AnchorPaneTest.java:35]: In our tests, we don't mix junit4 and junit5 APIs in single test class.
Could you please find an alternate/junit4 way for these two.

Comment (Maran23) [reply modules/javafx.graphics/src/test/java/test/javafx/scene/layout/AnchorPaneTest.java:35]: There is not really an alternative for this in junit4. I can create a new test class instead.

Comment (mstr2) [reply modules/javafx.graphics/src/test/java/test/javafx/scene/layout/AnchorPaneTest.java:35]: There's no need to do that. `AnchorPaneTest` works just as well if you simply use the JUnit5 imports exclusively.

Comment (Maran23) [reply modules/javafx.graphics/src/test/java/test/javafx/scene/layout/AnchorPaneTest.java:35]: good point, that will be the better solution.

**Comment by kevinrushforth** [comment]
NOTE: this is blocked by the fix for [JDK-8296283](https://bugs.openjdk.org/browse/JDK-8296283), which @arapte plans to do next week.

**Review by arapte — APPROVED** [review]
Looks good to me.

File: modules/javafx.graphics/src/test/java/test/javafx/scene/layout/AnchorPaneTest.java [hunk 2]
```diff
@@ -718,4 +725,56 @@ public void testBottomAnchoredMinSizeOverridden() {
...
+        Scene scene = new Scene(anchorpane, widthHeight, widthHeight);
+        stage.setScene(scene);
+        stage.show();
+
+        Bounds boundsInParent = child.getBoundsInParent();
+
+        double snappedPaddingX = child.snapPositionX(padding);
+        double snappedPaddingY = child.snapPositionY(padding);
+
+        assertEquals(snappedPaddingX, boundsInParent.getMinX(), 0.0001);
```
[the remark this answers is not in the document]
Comment (Maran23) [reply modules/javafx.graphics/src/test/java/test/javafx/scene/layout/AnchorPaneTest.java:765]: Not sure if this is worth it for just 2 occurrences. In general I always try not to clutter the test class with things used in only one or two test methods.

[the remark this answers is not in the document]
Comment (andy-goryachev-oracle) [reply modules/javafx.graphics/src/test/java/test/javafx/scene/layout/AnchorPaneTest.java:765]: I see 6 occurrences only in this method ;-)

[the remark this answers is not in the document]
Comment (Maran23) [reply modules/javafx.graphics/src/test/java/test/javafx/scene/layout/AnchorPaneTest.java:765]: ok :D still not sure if it is worth it for just one test.

File: modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java [hunk 3]
```diff
@@ -355,15 +350,15 @@ private double computeChildHeight(Node child, Double topAnchor, Double bottomAnc
             }
 
             if (leftAnchor != null) {
-                x = insets.getLeft() + leftAnchor;
+                x = snappedLeftInset() + leftAnchor;
```
[the remark this answers is not in the document]
Comment (andy-goryachev-oracle) [reply modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java:353]: Not sure about render scale - it should be done elsewhere.

As for snapping, I think this code might be incorrect - it should honor Region.snapToPixel setting.  All the intermediary computations need to use unsnapped (original) values, then, at the last moment, call proper Region.snap*() which either rounds or not, depending on the `snapToPixel`.

The only exception for this rule (of using snapped values in the intermediary computations) is to ensure some kind of specific alignment, which I believe is not the case here.

File: modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java [hunk 4]
```diff
@@ -355,15 +350,15 @@ private double computeChildHeight(Node child, Double topAnchor, Double bottomAnc
...
 
             if (topAnchor != null) {
-                y = insets.getTop() + topAnchor;
+                y = snappedTopInset() + topAnchor;
             } else if (bottomAnchor != null) {
-                y = getHeight() - insets.getBottom() - bottomAnchor - h;
+                y = getHeight() - snappedBottomInset() - bottomAnchor - h;
             }
 
             child.resizeRelocate(x, y, w, h);
```
Comment (andy-goryachev-oracle) [inline modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java:364]: snapping should be done at this moment, I think, something like this:
```
child.resizeRelocate(
  snapPositionX(x),
  snapPositionY(y),
  snapSizeX(w),
  snapSizeY(h));
```
and keep the original (unsnapped) code prior to this.

File: modules/javafx.graphics/src/test/java/test/javafx/scene/layout/AnchorPaneTest.java [hunk 5]
```diff
@@ -718,4 +725,56 @@ public void testBottomAnchoredMinSizeOverridden() {
...
+     * @see <a href="https://bugs.openjdk.org/browse/JDK-8295078">JDK-8295078</a>
+     */
+    @ValueSource(doubles = { 1.0, 1.25, 1.5, 1.75, 2.0 })
+    @ParameterizedTest
+    void testAnchorPaneSnappingWithDifferentScales(double scale) {
+        double padding = 9.6;
+
+        StackPane child = new StackPane();
+        AnchorPane anchorpane = new AnchorPane(child);
+        anchorpane.setStyle("-fx-padding: " + padding + "px;");
```
Comment (andy-goryachev-oracle) [inline modules/javafx.graphics/src/test/java/test/javafx/scene/layout/AnchorPaneTest.java:742]: There might be two cases: when snapToPixel=true (this is the default case), and when =false.  Should we test the case snapToPixel=false?

**Review by andy-goryachev-oracle — CHANGES_REQUESTED** [review]

File: modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Maran23) [reply modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java:353]: I'm not sure about this. Other places do the same, and there are lot of places actually which do something like `snappedLeftInsets` + `snappedRightInsets` or `final double w = snapSizeX(getWidth()) - x - snappedRightInset();` or similar. 
I think this was for a reason. 
And I have only oriented myself to the existing code in this case. If you are right, then this was done wrong everywhere.

Btw. I had a look and other container do snap their constraints, e.g. `H/VBox` its spacing, `GridPane` his `H/VGap`. So pretty sure we also should snap the anchor values.

File: modules/javafx.graphics/src/test/java/test/javafx/scene/layout/AnchorPaneTest.java (hunk 5 above)
Comment (Maran23) [reply modules/javafx.graphics/src/test/java/test/javafx/scene/layout/AnchorPaneTest.java:742]: We can, but that should do the same as before. As before the values were not snapped at all. And that worked, I guess 😄

File: modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (andy-goryachev-oracle) [reply modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java:353]: I think the idea is this (and it only applies when snapToPixels=true).  We only need to use snapped coordinates/sizes when placing components.  In order to avoid accumulating errors, which can be seen during gradual resizing for example, we should avoid snapping of intermediary results, UNLESS there is a possibility of introducing visual artifacts.

What I mean by this is imagine a container that aligns its children in a table-like layout (similar to https://github.com/andy-goryachev/FxDock/blob/master/doc/CPane.md).  Depending on constraints, there might be a situation where rounding might shift certain children in relation to the vertical guide lines - we don't want that.

So in this case we need to snap the guide lines as an intermediate step (i.e. compute the first guide line position, then compute the next column, snap, next column, etc.)

Whereas in a simple case (one child node), we can simply use snapped*Inset() - we don't have intermediate computations.

In the case of AnchorPane, I think, we should first sum insets and anchor, then snap - because developers might want to use or not to use snapToPixels, and might set a fractional anchor.

[the remark this answers is not in the document]
Comment (kevinrushforth) [reply modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java:353]: I initially had the same thought that Andy had, but I want to take a closer look -- especially at what similar layout panes do.

As for the point about doing this only when `snapPixels == true`, yes, but you don't need to worry about that if you call the `snapXXX` methods in Region. All of them check the `snapToPixel` attribute.

File: modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java [hunk 6]
```diff
@@ -280,8 +280,7 @@ private double computeWidth(final boolean minimum, final double height) {
                     child.minWidth(childHeight) : computeChildPrefAreaWidth(child, -1, null, childHeight, false)) + right);
         }
 
-        final Insets insets = getInsets();
-        return insets.getLeft() + max + insets.getRight();
+        return snappedLeftInset() + max + snappedRightInset();
```
Comment (kevinrushforth) [inline modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java:283]: Have you verified that `max` ends up as a snapped value in all cases? Some of the values that are used in the computation of max clearly are snapped, such as `computeChildPrefAreaWidth` (which is a method in the Region superclass that returns a snapped result), but others might not be.

File: modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (kevinrushforth) [reply modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java:353]: I looked at the other layout panes and they already do something similar to what this PR does, using the snapped insets for intermediate compuation. I also note that some of the computation is calling superclass methods that do snapping already, and expect it to be added to snapped insets (if there are insets), for example, `computeWidth()`. In this particular case -- `snappedLeftInset() + leftAnchor` -- I don't think `leftAnchor` is snapped, but it probably should be.

File: modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java (hunk 4 above)
Comment (kevinrushforth) [reply modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java:364]: It depends (see earlier discussion), but if we are going to snap the intermediate results, we need to be sure we are doing it consistently -- adding snapped insets to an unsnapped value is very likely the wrong thing to do.

**Review by kevinrushforth — COMMENTED** [review]
Not a complete review, but I did leave some inline comments. I note that there is precedent for using the snapped insets in intermediate computations, so that part of the change seems OK to me. If that is going to be the pattern, you need to make sure that you are snapping the values you add to the snapped insets.

I'd like @arapte to take a closer look as well.

File: modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Maran23) [reply modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java:353]: Yeah I think so too.

File: modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java (hunk 6 above)
Comment (Maran23) [reply modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java:283]: besides the anchor everything is snapped properly here already.

File: modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java (hunk 4 above)
Comment (andy-goryachev-oracle) [reply modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java:364]: not only for consistency, but we should also understand whether it makes sense or not.

for example, we might snap intermediate results such as guide lines, if that makes sense (e.g. for alignment), but most importantly, we must honor the snapToPixel setting for the end result.  If it's true, then all child nodes must have integer coordinates and sizes.

Would you agree?

**Comment by mstr2** [comment]
@andy-goryachev-oracle
Snapping doesn't necessarily imply integer coordinates or sizes. The reason is that we're snapping to _pixels_, not integers. For example, with 200% scale, every other pixel will have a .5 coordinate. It's also important to not confuse snapping with rounding to integers because integer coordinates might not even fall on pixel boundaries for some scaling factors.

**Comment by andy-goryachev-oracle** [comment]
@mstr2 : 
You are right, I wasn't aware of scale.  Snapping code looks like `Math.round(value * scale) / scale;`, so I take the 'whole integer' words back.  Sorry for the confusion.

Having said that, my earlier comment is still valid - if snapToPixel==true, the final location and sizes must go through snap*() methods.  And the earlier example is also valid, assuming scale=1.0.

File: modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Maran23) [reply modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java:353]: I also never understood that completely.
Hence I asked at the mailing list about snapping in the past: https://mail.openjdk.org/pipermail/openjfx-dev/2021-September/032166.html
In my opinion it also makes sense to snap the values once. But that would be a huge refactoring an everything is affected. We can not just change this now. Thats why I would stick to the existing solution and refactor that later.
But some clarification here would be helpful here, maybe ask the people who created this (if available)?. And I haven't looked yet, but how does Swing handle this?

[the remark this answers is not in the document]
Comment (arapte) [reply modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java:353]: It's seems little tricky to choose an approach for this fix.

- NOT snapping intermediate values (new approach):
1. It definitely seems correct NOT to snap intermediate values.
2. But currently, code Controls snap the intermediate values,
3. This PR seems to be the first change of snapping in a layout:
=> Will it regress if we have different approach in layout than in control ?
I think No, It won’t regress. @Maran23 Can you please try it with this fix ?
In that case, we can adapt the new way for layout and file a new JBS to refactor existing code.

**Comment by Maran23** [comment]
I thought about this one more time.
Example:
-  render scale of 1.25 (125%).
- `Control` with a width of 200px
- `left inset` = 10, `right inset` = 15
- `left anchor` = 10, `right anchor` = 15

Now the UI would look like this currently:
[`Left inset` = 10.4][`Left anchor` = 10.4][`Control` = 148.8][`Right anchor` = 15.2][`Right inset` = 15.2]
Control width is calculated like this: 200 - 10.4 - 15.2 - 10.4 - 15.2 = 148.8

If we would not snap intermediate values, the left side (inset + anchor) would be 20 instead of 20.8.
The right side would still be 30.4.

Thinking about this again, isn't the current behaviour what we would expect from the layout with a scale of 1.25?
Snapping 'intermediate' values seems logical to me (current vehaviour), but maybe I'm missing something.
But we should avoid snapping things more than one time.

File: modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (andy-goryachev-oracle) [reply modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java:385]: I think I now understand the snapping better, especially after discussion in https://github.com/openjdk/jfx/pull/1111

it looks like we should be operating in terms of final pixels, that is, in snapped coordinates.  this might mean what we should not, generally, mix snapped and unsnapped values:

result = snap(a) + snap(b); // ok

result = snap(snap(a) + b) // also ok

result = snap(snap(a) + b + c) // not ok, should be written as
result = snap(snap(a) + snap(b) + snap(c)) or
result = snap(a) + snap(b) + snap(c)

What do you think?

[the remark this answers is not in the document]
Comment (Maran23) [reply modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java:385]: I agree but that is also what I do here, isn't it?

File: modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java [hunk 7]
```diff
@@ -265,72 +265,99 @@ private double computeWidth(final boolean minimum, final double height) {
...
-        final Insets insets = getInsets();
-        return insets.getTop() + max + insets.getBottom();
+        return snappedTopInset() + max + snappedBottomInset();
     }
 
     private double computeChildWidth(Node child, Double leftAnchor, Double rightAnchor, double areaWidth, double height) {
         if (leftAnchor != null && rightAnchor != null && child.isResizable()) {
-            final Insets insets = getInsets();
-            return areaWidth - insets.getLeft() - insets.getRight() - leftAnchor - rightAnchor;
+            return areaWidth - snappedLeftInset() - snappedRightInset() - leftAnchor - rightAnchor;
```
[the remark this answers is not in the document]
Comment (andy-goryachev-oracle) [reply modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java:342]: it is not immediately clear that left/right anchors are snapped.
areaWidth, however, is not.
should it be snappedSize?

File: modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java [hunk 8]
```diff
@@ -265,72 +265,99 @@ private double computeWidth(final boolean minimum, final double height) {
...
         }
         return computeChildPrefAreaWidth(child, -1, Insets.EMPTY, height, true);
     }
 
     private double computeChildHeight(Node child, Double topAnchor, Double bottomAnchor, double areaHeight, double width) {
         if (topAnchor != null && bottomAnchor != null && child.isResizable()) {
-            final Insets insets = getInsets();
-            return areaHeight - insets.getTop() - insets.getBottom() - topAnchor - bottomAnchor;
+            return areaHeight - snappedTopInset() - snappedBottomInset() - topAnchor - bottomAnchor;
```
Comment (andy-goryachev-oracle) [inline modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java:349]: same comment about areaHeight - should it be snapSize'd?

File: modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (andy-goryachev-oracle) [reply modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java:385]: here: yes!

File: modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java (hunk 8 above)
Comment (Maran23) [reply modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java:349]: Well, given that we (should) snap everything including the calculation of the width, I expected that the `areaWidth` and `height`, which are basically `getWidth` and `getHeight` are already snapped. 
That was at least my thinking back then, not sure now. I need to check everything again since this was a while ago. :)

Comment (hjohn) [reply modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java:349]: I think once there is an unsnapped parent, all children should probably also be unsnapped (ie. the property should work like the visible property; if I turn it off for one parent, all children become invisible/unsnapped).

There would otherwise be little point in setting a parent to unsnapped if one didn't also turn off snapping on all the children.

Note that we tested what happens when you put snapped content in an unsnapped container; currently this results in all children to be misaligned (not pixel aligned). I think that's the only real solution, as trying to resnap your positions when your parent container is misaligning you would otherwise need to introduce some kind of flexible border that "absorbs" some of the misalignment so the rest of the content can be snapped again...

Basically, there are three strategies for children when dealing with an unsnapped parent:

- Just do snapping as normal (JavaFX currently) and when the parent translation is added, all final positions will be shifted by an unsnapped amount resulting in all calculations (even though they used snapping logic) to be in the "wrong" positions
- Ignore snapping when a parent is unsnapped and use standard calculations
- Somehow try to resnap content taking the parent misalignment into account; this is going to cause visual artifacts as the parent is moved, and requires some kind of flexible border

Comment (hjohn) [reply modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java:349]: > In this case getWidth() may not be snapped since the titled pane has been laid out by an unsnapped parent.

And when that happens, you can't do anything about it either.  If you adjust the width, you risk not covering the entire area that the parent gave you (or more if you round up), and if you use it in any calculations you need to resnap any values resulting from it, and you would still end up with some dead space that your snapped calculations can't cover...

The only real solution seems to be to treat snapping like visibility IMHO.

File: modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (hjohn) [reply modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java:385]: This looks good to me, the snapping as explained by Andy is how it should be.  A source of error can be still be that floating point calculations can introduce small deviations so sometimes it is needed to resnap after doing calculations that seem harmless (ie. `snappedSize * 10` or `snappedWidth - snappedLeftInset` may need resnapping).

**Comment by mstr2** [comment]
I think we should document the insights gained in this discussion (and others) somewhere in JavaFX, since snapping is such a difficult thing to get right, and the source of so many visual artifacts.

File: modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java (hunk 8 above)
Comment (kevinrushforth) [reply modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java:349]: > I think once there is an unsnapped parent, all children should probably also be unsnapped (ie. the property should work like the visible property; if I turn it off for one parent, all children become invisible/unsnapped).

One complication with this is that the `snapToPixel` property is on `Region`, so `Group` (or any other `Parent` that isn't a `Region`) doesn't have that property. What you describe _might_ be OK as long as you ignore any parent that isn't a `Region` when asking the question "are all my ancestors have `snapToPixel`".

> Basically, there are three strategies for children when dealing with an unsnapped parent:
> 
> * Just do snapping as normal (JavaFX currently) and when the parent translation is added, all final positions will be shifted by an unsnapped amount resulting in all calculations (even though they used snapping logic) to be in the "wrong" positions

Right. If a non-snap-to-pixel ancestor lays out it children, then all bets are off. Absent a compelling reason to change it, this seems like a reasonable default (perhaps it should be documented better).

> * Ignore snapping when a parent is unsnapped and use standard calculations

A change to do this _might_ be OK, with the caveat I mentioned above about non-Region parents, but this would need careful analysis. As part of this, I would want to understand the motivation for doing this (i.e., what is the benefit?).

> * Somehow try to resnap content taking the parent misalignment into account; this is going to cause visual artifacts as the parent is moved, and requires some kind of flexible border

Yeah, this doesn't seems like a good idea.

If you want to explore this further, you could file a new RFE to consider making a change.

Comment (andy-goryachev-oracle) [reply modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java:349]: I think it's probably not worth doing.  When the application developer sets snapping off when it's on by default, they do it for a reason, so the current behavior is the right one.  May be they want to smooth some transition during animation (a-la scrolling ticker tape), who knows. 

FX gives this choice at per-Region basis for this kind of flexibility, otherwise there would be one global setting, right?

Comment (hjohn) [reply modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java:349]: > I think it's probably not worth doing. When the application developer sets snapping off when it's on by default, they do it for a reason, so the current behavior is the right one. May be they want to smooth some transition during animation (a-la scrolling ticker tape), who knows.
> 
> FX gives this choice at per-Region basis for this kind of flexibility, otherwise there would be one global setting, right?

For sure it shouldn't be a global setting, and the per Region choice seems absolutely the right one. What I was talking about is more how sensible it would be for children of an unsnapped Region to still use snapping logic, and if it shouldn't work similar to how visibility works (ie. if a parent is invisible, all its children are invisible, regardless of those children's individual visible status). 

If the scrolling ticker tape you mentioned contains any other children, the user probably wants snapping off for those as well as the ticker tape would otherwise not be as smooth as it could be; if the parent region is misaligned, all the children (if snapped) would still advance by an exact multiple of pixels, but they would always fall in the same spot between pixels.  Also, if the speed is not a nice pixel multiple (ie. 4.5 pixels per frame) then the children would sometimes move 4 pixels, sometimes 5.

It's not truly a big issue, just something that may be unexpected.

Comment (hjohn) [reply modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java:349]: @kevinrushforth it probably isn't worth pursuing a change in the child/parent snapping logic right now, at least not without some compelling case. First I think snapping is rarely disabled (I never have, despite working on graphics heavy JavaFX applications).  Second, one can already get the (possibly more) desired situation by taking care all children are unsnapped as well.

Comment (Maran23) [reply modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java:349]: I agree with the consensus here. And same as @hjohn, I actually also never disabled snapping. There was never usecase where I needed to do so (And I worked on a lot of UI related things, including creating custom components of all kind).

> Right. If a non-snap-to-pixel ancestor lays out it children, then all bets are off. Absent a compelling reason to change it, this seems like a reasonable default (perhaps it should be documented better).

I agree here.The defaultis fine, and it's hard to say what would be a better default 'setting' here (if there really is a better way). Does a developer really expect that all children of a unsnapped node are unsnapped as well? Or is there a usecase where you don't want this? It's hard to answer, so as written above, the default seems fine and logicial.

File: modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Maran23) [reply modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java:385]: Good point. That is something we should think about - this may or may not be a problem with some of the snapping code.

**Review by kevinrushforth — APPROVED** [review]
